composition-js/src/merging/merge.ts

```diff
@@ -471,16 +490,35 @@ class Merger {
     // However, in practice and for "execution" directives, we will only keep the the ones
     // that are in _all_ subgraphs. But we're do the remove later, and while this is all a
     // bit round-about, it's a tad simpler code-wise to do this way.
+    const customDirectiveMap = new Map<string, DirectiveDefinition>();
     for (const subgraph of this.subgraphsSchema) {
       for (const directive of subgraph.allDirectives()) {
         if (!this.isMergedDirective(directive)) {
           continue;
         }
         if (!this.merged.directive(directive.name)) {
-          this.merged.addDirectiveDefinition(new DirectiveDefinition(directive.name));
+          const def = this.merged.addDirectiveDefinition(new DirectiveDefinition(directive.name));
+          if (this.exposedDirectives().includes(directive.name)) {
+            customDirectiveMap.set(directive.name, def);
+          }
         }
       }
     }
+
+    // now for all custom directives, make sure locations are properly added
+    customDirectiveMap.forEach((def, directiveName) => {
+      const definitions: DirectiveDefinition[] = this.subgraphsSchema
+        .map(schema => schema.directive(directiveName))
+        .filter((def): def is DirectiveDefinition => def !== undefined);
+      // if any directive definition has a executable location, we'll just add all locations
+      // but if it's a type-system only directive, we can attempt to merge the directives together
+      if (definitions.some(d => d.locations.some(loc => executableDirectiveLocations.includes(loc)))) {
```

Comment:
I'm not sure the new logic regarding executable-only  and "mixed" (executable and type system locations) directives.

In particular, it's confusing that we assign locations to the merged definition here, but do it again in `mergeExecutableDirectiveDefinition`. And here, we seems to assign any locations found in any subgraph, which is wrong for executable locations (we should only merge a given executable location if it is in _all_ subgraph; because we cannot forbid a user to include the directive in a query anywhere the merged locations permit, but if one of those application reach a subgraph that does not have the location, that subgraph query will break). So as far as I can tell, this break the proper merging of executable directives, and the fact that no unit test broke means we need to fix that coverage. 

More generally, the merging logic is that executable-only directives are merged by intersection, while type-system-only ones are merged by union. But I think that means mixed ones needs a middle ground strategy. More precisely, I think that for mixed directives:
* we should intersect executable locations but union type-system locations. As said above, not intersecting executable locations would break subgraph queries down the line, and not unioning type-system locations would result in invalid supergraphs.
* we should hard error out if `repeatable` is not consistent. If it's inconsistent, all choices in the subgraph will create issues: if we mark it repeatable, then subgraph query will break a query with the directive repeatable reach a subgraph where the definition is not repeatable, but if we mark it non-repeatable, we may have some repeated applications in some subgraph and upon merging, the supergraph will be invalid graphQL.

In other words, I don't think we can treat mixed directives the same as either executable-only or type-system only  ones.

Reply:
I've checked in the code for the first half of this. Will come back to the second part tomorrow.